mem/buffers.go

```diff
@@ -0,0 +1,288 @@
+func (b *Buffer) Len() int {
+	// Convenience: io.Reader returns (n int, err error), and n is often checked
+	// before err is checked. To mimic this, Len() should work on nil Buffers.
+	if b == nil {
+		return 0
+	}
+	return len(b.ReadOnlyData())
+}
+
+// Split creates a new reference to the first n bytes of the underlying buffer,
+// and modifies the receiver to only point to the remaining bytes. The returned
+// Buffer functions just like a normal reference acquired using Ref in that the
+// underlying buffer will only be freed once both the receiver and the new Buffer
+// are freed.
+func (b *Buffer) Split(n int) *Buffer {
+	if b.freed {
+		panic("Cannot split freed buffer")
+	}
+
+	b.refs.Add(1)
+
+	data := b.data
+	var free func([]byte)
+	if f := b.free; f != nil {
+		free = func(_ []byte) {
+			f(data)
+		}
+	}
+
+	b.data = data[:n]
+	b.free = nil
+
+	return &Buffer{
+		data: data[n:],
+		refs: b.refs,
+		free: free,
+	}
+}
+
+func (b *Buffer) String() string {
+	return fmt.Sprintf("mem.Buffer(%p, data: %p)", b, b.ReadOnlyData())
+}
+
+// BufferSlice offers a means to represent data that spans one or more Buffer
+// instances.
+type BufferSlice []*Buffer
+
+// Len returns the sum of the length of all the Buffers in this slice. Warning:
+// invoking the built-in len on a BufferSlice will return the number of buffers
+// in the slice, and *not* the value returned by this function.
+func (s BufferSlice) Len() (length int) {
+	for _, b := range s {
+		length += b.Len()
+	}
+	return length
+}
+
+// Ref returns a new BufferSlice containing a new reference of each Buffer in the
+// input slice.
+func (s BufferSlice) Ref() BufferSlice {
+	out := make(BufferSlice, len(s))
+	for i, b := range s {
+		out[i] = b.Ref()
+	}
+	return out
+}
+
+// Free invokes Buffer.Free on each Buffer in the slice.
+func (s BufferSlice) Free() {
+	for _, b := range s {
+		b.Free()
+	}
+}
+
+// CopyTo copies each of the underlying Buffer's data into the given buffer,
+// returning the number of bytes copied.
+func (s BufferSlice) CopyTo(out []byte) int {
+	off := 0
+	for _, b := range s {
+		off += copy(out[off:], b.ReadOnlyData())
+	}
+	return off
+}
+
+// Materialize concatenates all the underlying Buffer's data into a single
+// contiguous buffer using CopyTo.
+func (s BufferSlice) Materialize() []byte {
+	out := make([]byte, s.Len())
+	s.CopyTo(out)
+	return out
+}
+
+// LazyMaterialize functions like Materialize except that it writes the data to a
+// single Buffer pulled from the given BufferPool. As a special case, if the
+// input BufferSlice only actually has one Buffer, this function has nothing to
+// do and simply returns said Buffer, hence it being "lazy".
+func (s BufferSlice) LazyMaterialize(pool BufferPool) *Buffer {
+	if len(s) == 1 {
+		return s[0].Ref()
+	}
+	buf := pool.Get(s.Len())
+	s.CopyTo(buf)
+	return NewBuffer(buf, pool.Put)
+}
+
+// Reader returns a new Reader for the input slice.
+func (s BufferSlice) Reader() *Reader {
+	return &Reader{
```

Comment:
And if we can throw in a `Close` method on this `Reader` that frees up any unused buffers, we can make it implemented `io.ReadCloser` as well. Thoughts?

Reply:
ACK I can add that documentation. There's a specific reason for that, which is that `Reader` is used specifically in the `loopyWriter`, which can fail partially through the write and be retried later. This means freeing the buffers as the reader goes through them would make it impossible to retry. I can definitely add `Close` on it though, since it would remove needing to keep the reference to the original slice around in [`dataFrame`](https://github.com/PapaCharlie/grpc-go/blob/017b72fe45c6a1960468887642a0fc2f22771430/internal/transport/controlbuf.go#L152).